How-To: Monitor User Workloads using Blackbox Exporter (#137)
appuio/appuio-cloud-docs: docs/modules/ROOT/pages/how-to/monitor-http-endpoints.adoc

= Monitor HTTP(S) Endpoints Using Blackbox Exporter

== Prerequisites

* Command Line Access to APPUiO Cloud
* Helm 3, https://helm.sh/

== Install Blackbox Exporter

. Set the namespace for the Blackbox Exporter
+
[source,bash]
----
APP_NAMESPACE="my-app" <1>
EXPORTER_NAMESPACE="my-org-blackbox-exporter" <2>
----
<1> The namespace of your application.
<2> The exporter can be installed in the same namespace as your application or in a separate namespace.

> You might need additional network policies if installing the exporter in a separate namespace.

Reviewer: ```suggestion
If you deploy the Blackbox Exporter in a separate namespace and want to use in-cluster probes, you'll have to configure additional network policies as the default network policies configured by {product} isolate namepaces from each other.
```

> You might need additional network policies if installing the exporter in a separate namespace.

Author: That's an oversight i wanted the default to be the same namespace

```
[source,bash]
----
APP_NAMESPACE="my-app" <1>
EXPORTER_NAMESPACE="${APP_NAMESPACE}" <2>
----

```

> See the <<tips-and-tricks>> section for more information.

Reviewer: There's no networkpolicy shown in tips-and-tricks. So there's no way to end up with a working setup by following this guide using the default setup of having the blackbox-exporter in a separate namespace.

Query all available metrics with `{job=~"probe/.*"}`.
<2> The `for` clause specifies how long the probe has to fail before the alert is triggered.
Shorter times might help find problems earlier but might also trigger false positives or flukes.

. Verify the alert was picked up by Prometheus
+
Navigate to the OpenShift Web Console https://console.<CLUSTER>.cloud/dev-monitoring/ns/<APP_NAMESPACE>/alerts.
+
image::monitoring/blackbox-alert-list.png[List of Blackbox alerts https://console.<CLUSTER>.cloud/dev-monitoring/ns/<APP_NAMESPACE>/alerts]
+
The alert should show up in the list.
If it doesn't show up see the <<troubleshooting>> section.

== Troubleshooting [[troubleshooting]]

=== No Entry is Shown in the List if Querying the `up` Metric `up{job=~"probe/.*"}`

[INFO]
The probe CRD isn't well-formed or Prometheus hasn't yet scraped the probe.

Prometheus hasn't yet scraped the probe.
Wait for three more minutes and check again.

If no entry is shown in the list, the `Probe` CRD is most likely is misconfigured.
Especially check the `spec.prober.url`.
The format is `HOST:PORT`.
`http://HOST` isn't valid.

=== The Value of the `up` Metric is `0`

[INFO]

> The probe CRD is well-formed but the blackbox exporter isn't reachable.

Reviewer: Above, you've capitalized "Blackbox Exporter", but starting here you don't anymore. Please pick one variant and stick to it.

If the value is `0`, the `Probe` CRD is valid, but the blackbox exporter can't be reached.
Double check the `.spec.prober.url` of the `Probe` CRD.

Check if the blackbox exporters deployment is running and shown as ready.

=== `probe_success` shows `0`

[INFO]
The probe CRD is well-formed and the blackbox exporter is reachable.
The target isn't reachable, the certificate is bad, or the target returns an unwanted status code.

Check the `probe_http_status_code` for the job.
Only `2XX` is considered a success after the final redirect.
If the status code is `0`, the problem is earlier in the chain.

=== No Alerts Shows Up in the Alert List

// The promtheus rule is validated by the prometheus operator. I did not manage to apply a rule that was not valid.

[INFO]
The `PrometheusRule` CRD might not be well-formed and might have failed to apply.

. Check if the `PrometheusRule` CRD was created successfully.
+
[source,shell]
----
kubectl -n "${APP_NAMESPACE}" get prometheusrules
----

> === Debug Non-Successful Probes

Reviewer: Non-Successful is a mouthful, I'd suggest "Failing" instead.

```suggestion
=== Debug Failing Probes
```

. Connect to the blackbox exporter
+
[source,bash]
----
kubectl --namespace "${EXPORTER_NAMESPACE}" port-forward svc/blackbox-exporter-prometheus-blackbox-exporter 9115
----

> . Open the blackbox exporter in a browser

Reviewer: Genuine question: Can the blackbox exporter probe logs be checked with `kubectl logs` at all?

Author: Not in the default configuration at least. Should i invest some more time here to check if it can be enabled?

Reviewer: I wouldn't spend too much time, but it might be worth it to check if there's an obvious option in blackbox exporter options.

+
Navigate to http://localhost:9115.

. Choose the failed target from the list
+
image::monitoring/blackbox-probe-debug.png[Blackbox exporter debug interface http://localhost:9115]
+
Click __logs__ on the failed target.

. Check the logs
+
The logs should show the reason for the failure.
In this example an expired certificate is the cause of the error.
+
image::monitoring/blackbox-probe-debug-logs.png[Blackbox exporter debug logs http://localhost:9115/logs?id=<ID>]

== Tips & Tricks [[tips-and-tricks]]

=== Public (Ingress) vs. Local Probes

Local probes have a higher chance of just monitoring the application.
They don't have to go through the ingress so you won't be notified if there are problems with the ingress.
However, it isn't possible to detect expired certificates or misconfigured ingresses.

Public probes have a higher chance of detecting false positives.

> For example the application works but the ingress is down.

Reviewer: ```suggestion
For example, a public probe will fail if the application works, but the ingress is down.
```

> But they allow you to detect expired certificates and misconfigured ingresses.

Reviewer: ```suggestion
However, they allow you to setup alerts for expiring certificates configured on the application's ingress and misconfigurations of the ingress itself.
```

=== Cross Cluster Probes

The blackbox exporter can be installed on a different zone or cluster than the application and probe the public endpoint of the application.

[WARNING]
There is a big unknown (the internet) in between clusters.
There is a high chance that the probe will fail because of unrelated network issues between the two clusters.

=== TLS Checks

The blackbox exporter returns metrics for the TLS certificate validity period.

> It's thus possible to alert on expired certificates with the `probe_ssl_earliest_cert_expiry` metric.

Reviewer: ```suggestion
Using this information, you can alert on expired certificates by checking the `probe_ssl_earliest_cert_expiry` metric.
```

[source,yaml]
----
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  name: my-app-blackbox-alerts
spec:
  groups:
  - name: my-app-blackbox.rules
    rules:
    - alert: TLSCertExpiringSoon
      expr: probe_ssl_earliest_cert_expiry - time() < 5 * 86400 <1>
      for: 1m
      labels:
        severity: warning
      annotations:
        description: "TLS certificate for {{ $labels.instance }} will expire in {{ $value | humanizeDuration }}"
    - alert: EndpointDown
      expr: probe_success == 0
      for: 5m
      labels:
        severity: critical
      annotations:
        summary: "Endpoint {{ $labels.instance }} down"
----
<1> The `probe_ssl_earliest_cert_expiry` metric is the earliest expiration date of the certificate chain.

> 5 days are chosen to give enough time to renew the certificate even through extended weekends.

Reviewer: ```suggestion
5 days are chosen to give enough time to renew the certificate even through extended weekends.

TIP: If you're using Let's Encrypt certificates, you may want to alert for certificates expiring in 20 days or less, as the default configuration for {product} renews Let's Encrypt certificates when their remaining validity is 30 days.
```
